File: DSA/NiharikaChand_590017697/Day22Question1.java
import java.util.*;
class Day22Question1 
{
    static class Node 
    {
        int data;
        Node next;

        Node(int data) 
        {
            this.data = data;
            this.next = null;
        }
    }
    Node head;

    public void insert(int data) 
    {
        Node newNode = new Node(data);
        if (head == null) 
        {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) 
        {
            temp = temp.next;
        }
        temp.next = newNode;
    }

    public void findMid() 
    {
        if (head == null) 
        {
            System.out.println("List is empty");
        } 
        else 
        {
            Node temp = head;
            int c = 0;

            while (temp != null) 
            {
                c++;
                temp = temp.next;
            }
            int midPos = c / 2;
            temp = head; 
            int i;
            for (i = 0; i < midPos; i++)
            {
                temp = temp.next;
            }
            System.out.println("Middle element: " + temp.data);
        }
    }

    public static void main(String args[]) 
    {
        Day22Question1 list = new Day22Question1();
        list.insert(2);
        list.insert(4);
        list.insert(6);
        list.insert(7);
        list.insert(5);
        list.insert(1);

        list.findMid();
    }
}

File: DSA/PriyankaSingh_590016827/day22question1.c
#include <stdio.h>
#include <stdlib.h>

// Define structure for a node
struct Node {
    int data;
    struct Node* next;
};

// Function to create a new node
struct Node* createNode(int value) 
{
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = NULL;
    return newNode;
}

// Function to insert node at the end
void insertEnd(struct Node** head, int value) 
{
    struct Node* newNode = createNode(value);
    if (*head == NULL) 
    {
        *head = newNode;
        return;
    }
    struct Node* temp = *head;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
}

// Function to find the middle node using two-pointer approach
void findMiddle(struct Node* head) 
{
    struct Node* slow = head;
    struct Node* fast = head;

    while (fast != NULL && fast->next != NULL) 
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    if (slow != NULL)
        printf("Middle node: %d\n", slow->data);
}

// Function to print the list (optional)
void printList(struct Node* head) 
{
    while (head != NULL) 
    {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main() 
{
    struct Node* head = NULL;
    int n, value;

    printf("Enter number of elements in the linked list: ");
    scanf("%d", &n);

    printf("Enter %d elements:\n", n);
    for (int i = 0; i < n; i++) 
    {
        scanf("%d", &value);
        insertEnd(&head, value);
    }

    printf("Linked List: ");
    printList(head);

    findMiddle(head);

    return 0;
}
